src/backend/distributed/commands/foreign_constraint.c

```diff
@@ -696,89 +618,154 @@ HeapTupleOfForeignConstraintIncludesColumn(HeapTuple heapTuple, Oid relationId,
+	List *foreignKeyOids = GetForeignKeyOidsInternal(relationId, flags);
 
-	Relation pgConstraint = heap_open(ConstraintRelationId, AccessShareLock);
+	return list_length(foreignKeyOids) > 0;
+}
 
-	ScanKeyInit(&scanKey[0], Anum_pg_constraint_conrelid, BTEqualStrategyNumber, F_OIDEQ,
-				relationId);
-	SysScanDesc scanDescriptor = systable_beginscan(pgConstraint, scanIndexId, useIndex,
-													NULL,
-													scanKeyCount, scanKey);
 
-	HeapTuple heapTuple = systable_getnext(scanDescriptor);
-	while (HeapTupleIsValid(heapTuple))
+/*
+ * ConstraintIsAForeignKey returns true if the given constraint name
+ * is a foreign key defined on the relation.
+ */
+bool
+ConstraintIsAForeignKey(char *inputConstaintName, Oid relationId)
+{
+	int flags = EXTRACT_REFERENCING_CONSTRAINTS;
+	List *foreignKeyOids = GetForeignKeyOidsInternal(relationId, flags);
+
+	Oid foreignKeyOid = FindForeignKeyOidWithName(foreignKeyOids, inputConstaintName);
+
+	return OidIsValid(foreignKeyOid);
+}
+
+
+/*
+ * FindForeignKeyOidWithName searches the foreign key constraint with
+ * inputConstraintName in the given list of foreign key constraint OIDs.
+ * Returns the OID of the matching constraint. If there no matching constraint
+ * in the given list, then returns InvalidOid.
+ */
+static Oid
+FindForeignKeyOidWithName(List *foreignKeyOids, const char *inputConstraintName)
+{
+	Oid foreignKeyOid = InvalidOid;
+	foreach_oid(foreignKeyOid, foreignKeyOids)
 	{
-		Form_pg_constraint constraintForm = (Form_pg_constraint) GETSTRUCT(heapTuple);
+		char *constraintName = get_constraint_name(foreignKeyOid);
 
-		if (constraintForm->contype == CONSTRAINT_FOREIGN)
-		{
-			systable_endscan(scanDescriptor);
-			heap_close(pgConstraint, NoLock);
+		Assert(constraintName != NULL);
 
-			return true;
+		if (strncmp(constraintName, inputConstraintName, NAMEDATALEN) == 0)
+		{
+			return foreignKeyOid;
 		}
-
-		heapTuple = systable_getnext(scanDescriptor);
 	}
 
-	systable_endscan(scanDescriptor);
-	heap_close(pgConstraint, NoLock);
-
-	return false;
+	return InvalidOid;
 }
 
 
 /*
- * ConstraintIsAForeignKey returns true if the given constraint name
- * is a foreign key to defined on the relation.
+ * GetForeignKeyOidsInternal takes in a relationId, and returns
+ * the list of OIDs for foreign constraints that the relation with relationId
+ * is involved according to "flags" argument. See ExtractForeignKeyConstrainstMode
+ * enum definition for usage of the flags.
  */
-bool
-ConstraintIsAForeignKey(char *constraintNameInput, Oid relationId)
+static List *
+GetForeignKeyOidsInternal(Oid relationId, int flags)
 {
+	AttrNumber pgConstraintTargetAttrNumber = InvalidAttrNumber;
+
+	/*
+	 * Only one of them should be passed at a time since the way we scan
+	 * pg_constraint differs for those columns. Anum_pg_constraint_conrelid
+	 * supports index scan while Anum_pg_constraint_confrelid does not.
+	 */
+	Assert(!((flags & EXTRACT_REFERENCING_CONSTRAINTS) &&
```

Comment:
This one is useful to detect if both is passed. Should we have another assert that at least one is passed?
```C
bool extractReferencing PG_USED_ASSERT_ONLY = (flags & EXTRACT_REFERENCING_CONSTRAINTS);
bool extractReferenced PG_USED_ASSERT_ONLY = (flags & EXTRACT_REFERENCED_CONSTRAINTS);
 
/* use the same comment as above */
Assert (! (extractReferencing && extractReferenced));

/* at least one should be passed */
Assert (extractReferencing || extractReferenced))